db/migrate/20210519204247_move_miq_set_memberships.rb

```diff
@@ -0,0 +1,65 @@
+class MoveMiqSetMemberships < ActiveRecord::Migration[5.0]
+  class Relationship < ActiveRecord::Base
+    self.inheritance_column = :_type_disabled # disable STI
+  end
+
+  class MiqSetMembership < ActiveRecord::Base
+  end
+
+  def up
+    say_with_time("Converting set relationships to miq_set_memberships") do
+      # The `ancestry` column holds the id for the `relationship` record that
+      # references the miq_set the parent record contains the miq_set object
+      #
+      # Because these are trivial trees (1 parent with 1 level of children),
+      # the `miq_set_id` can be derived from the `resource_id` of the parent
+      miq_set_id = "(SELECT rel2.resource_id FROM relationships rel2 WHERE rel2.id = rels.ancestry::bigint)"
+
+      connection.execute(<<-SQL.squish)
+        INSERT INTO miq_set_memberships (member_type, miq_set_id, member_id, created_at, updated_at)
+          SELECT rels.resource_type, #{miq_set_id}, rels.resource_id, rels.created_at, rels.updated_at
+          FROM relationships rels
+          WHERE rels.relationship = 'membership' AND rels.ancestry IS NOT NULL
+      SQL
```

Comment:
Maybe it's just me, but I find something like the following much more readable:

```ruby
miq_set_ids = Relationship.where(:relationship => "membership", :ancestry => nil).pluck(:id, :resource_id).to_h

Relationship.where(:relationship => "membership").where.not(:ancestry => nil).each do |r|
  MiqSetMembership.create!(
    :miq_set_id  => miq_set_ids[r.ancestry.to_i],
    :member_type => r.resource_type,
    :member_id   => r.resource_id,
    :created_at  => r.created_at,
    :updated_at  => r.updated_at
  )
end
```

Reply:
Yes, but it also is a major N+1, and on large datasets I would prefer to make this a single statement where possible.

Reply:
There are unlikely no large datasets, so while it is technically an N+1, it will have no practical difference.  See also my comment here (though I admittedly hedged there on us looking at some "real" databases): https://github.com/ManageIQ/manageiq-schema/pull/536#discussion_r524354668

Reply:
Interesting, I just noticed the original code in #536 was AR based, and was switched _to_ pure SQL...not sure why.

Reply:
Two things with this:

- I was responding to this comment https://github.com/ManageIQ/manageiq-schema/pull/536#discussion_r530650239 and trying to make it work like the `:up`
- In that form **wrong** because it only re-added the children, and not the parent nodes.  The switch to `miq_set_members` flattens the relationships, so the `:down` needs to put it back in a tree, which it wasn't doing previously.

I wasn't in the business of making this a complete rewrite, just getting it to where it worked properly.

(P.S.  This is also code you originally wrote)

Reply:
Yeah I'm ok with merging as is if the tests all pass, I just find pure SQL really hard to read

Reply:
for this one, they read the same to me.
But for my test database, it took tens of seconds (possibly in the minutes?) vs 100ms